< [1 The Basics of Python](1-TheBasics.ipynb) | [Contents](0-Contents.ipynb) | [3 Control Structures](3-ControlStructures.ipynb) >

# 2 Basic Datatypes
## 2.1 Basic datatypes
There are a few basic datatypes in Python. We already encountered two of them: integers (abbreviated `int`) and decimal numbers (abbreviated `float`). But there are more.

#### 2.1.1 Logical values
A third datatype is boolean (or logical, abbreviated `bool`). When you **compare** numbers, strings, etc. the result is either `True` (1) or `False` (0). Check the following examples:

In [ ]:
5 > -5

In [ ]:
6**2 < 2**5

#### 2.1.2 Strings
Another datatype is string (abbreviated `str`). We already met strings in the previous chapter as the variable `s = "The time is 13:24."`

A string is a combination of letters and numbers and other characters (punctuation marks, special characters, etc.). Whenever you place letters and numbers in between (single or double) quotes you get a string. 

Have a look at the following examples (execute the code cells):

In [1]:
"This is a string with 7 words!"

'This is a string with 7 words!'

In [2]:
"9876"

'9876'

In [3]:
"3.14159"

'3.14159'

Note that even though we typed only numbers when entering 9876 we still have a string because of the double quotes.

To check the datatype we can use the built in `type()` function:

In [4]:
type("9876")

str

Note the difference between this and the next example.

In [5]:
type(9876)

int

In [6]:
type(3.14159)

float

In [7]:
type(False)

bool

#### 2.2.3 String concatenation and duplication
With the `+` and `*` operators strings can be **concatened** and **duplicated**.

In [ ]:
# giving variables text, and adding text.
word1 = "Good"
word2 = "morning"
word3 = "to you too!"
# string concatenation with the print() function
print(word1, word2)

# string concatenation with the + operator
sentence1 = word1 + word2 + word3 
print(sentence1)
sentence2 = word1 + " " + word2 + " " + word3 
print(sentence2)

# string duplication with the * operator
sentence3 = 3*word1
print(sentence3)
print("-"*12)

We observe that:
* the `print()` adds one space in between strings
* with the `+` operator strings are concatenated without extra spaces in between
* string duplication with the `*` operator saves typing time

The `len()` function gives the **number of characters** in a string.:

In [ ]:
text = "abc def, ghij!"
len(text)

Note that the spaces, the comma and the exclamation mark also take places: the 14 we obtained is the sum of
* 11 letters
* 2 spaces
* 1 exclamation mark 

#### 2.2.4 String manipulation
Often we need to manipulate strings:
* search for a character or a substring,
* count the numberof occurances,
* replace one or more characters by others,
* split the string,
* etc. 

Python has an extensive list of string methods. Some useful methods are:

| Method | Description |
| :---: | --- |
| `find()` | Searches the string for a specified value and returns the position of where it was found |
| `count()` | Returns the number of times a specified value occurs in a string |
| `replace()` | Returns a string where a specified value is replaced with a specified value |
| `split()` | Splits the string at the specified separator, and returns a list |
| `lower()` | Converts a string into **lower** case |
| `upper()` | Converts a string into **upper** case |

The following examples illustrate the use of these string methods:

In [ ]:
word1 = "Good"
word2 = "morning"
word3 = "to you too!"

print(word1.count("o"))

In [ ]:
print(word2.upper())

In [ ]:
print(word3.split(" "))

In [ ]:
print(word3.replace("too!", "my friend."))

#### 2.2.5 Accessing Elements in a string
In many applications we have to address a single character or make selections from text. Strings are similar to `lists` which we will introduce later. Similar operations (called *list slicing*) apply to strings.

#### Indexing
Each character in a string has an index. Indexing starts at 0:
* the first (position 1) character has index 0, 
* the second (position 2) character has index 1, etc. 

Access to a character is granted by the use of the **index operator** `[]`:

```Python
    s[i] # with i ranging from 0 to len(s)-1
```
means: select the character with index `i` (position `i+1`) from the string `s`. The index `i` ranges from `0` to `len(s)-1`. 

Examples:

In [ ]:
s = "abcdefghij"
print(s[0])    # FIRST character

In [ ]:
print(s[2])    # THIRD character

Note that index `2` refers to the character at position `3`.

Indexing with **negative indices** is also possible. The syntax is (`s` is a string):

```Python
    s[-i] # with i ranging from 1 to len(s)
```

In [ ]:
print(s[-2])

We have selected the second character from the right. Note that here we start counting with -1 for the first character on the right.

The last character of a string can be indexed with a negative index using `-1`, the first character has the negative index `-len(text)`:

In [ ]:
s = "abcdefghij"
print(s[-1])         # LAST character, equivalent with s[len(s)-1]

In [ ]:
print(s[-len(s)]) # FIRST character, equivalent with s[0]

#### Slicing
Slicing is the selection of multiple characters. Again, the index operator `[]` is used:

```Python
    s[i:j] # with i < j
```

means: select the characters with indices `i`, `i+1`, up to `j-1`. The character with index `j` is **not included** in the selection.

The **general** syntax is:

```python
    s[start:stop:step]
```
where 
* `start` is the index of the first element (default is 0), 
* `stop` is the index of the last element (**not included**, default is len(list_name)) and 
* `step` is the number of elements to skip (default is 1).

Example:

In [ ]:
print(s[0:4])   # select characters with indices 0 up to 3

The following table summarizes some indexing and slicing possibilities (the variable `s` represents a string):

| Slice | Meaning |
| :---: | --- |
| `s[i]` | character at index `i` (counting starts at 0) |
| `s[i:j]` | characters `i` up to `j-1` (character `j` **not** included) |
| `s[i:]` | characters from `i` up to end of string |
| `s[:j]` | characters from start up to `j-1` (character  `j` **not** included) |
| `s[i:j:k]` | characters from `i` up to `j-1` in steps of `k` |
| `s[:]` | all characters |
| `s[::-1]` | all characters in **reversed** order|

Important note: with ranges, the **stop value is not included in the selection**.

Some examples:

In [ ]:
s = "abcdefghij"
print(s[4:8])

Here we see that `[4:8]` selects characters with indices 4, 5, 6, and 7 which is "efgh". 

In [ ]:
print(s[:4])

In [ ]:
print(s[4:])

With `[4:]` we start with index 4 until the end of the string, which results in "efghij".

In [ ]:
print(s[1:7:3])

In the last example only the characters with indices 1 and 4 are selected. The character with index 7 is not included in the selection.

#### 2.1.3 Summary
Let us summarize the datatypes we already know in a table:

| Abbreviation | Datatype | Example(s) |
| :---: | --- | --- |
| `int` | integer | 5, -73 |
| `float` | decimal number | 3.14, -29.469 |
| `str` | string | "word", "12.3" |
| `bool` | boolean | `True` (1) and `False` (0) |

#### Exercise 1

Write code below (replace the ... in the last command) to select the **second word** of the sentence using the correct **range**.

In [ ]:
word1 = "Good"
word2 = "morning"
word3 = "to you too!"
sentence1 = word1 + " " + word2 + " " + word3
print(sentence1)
secondWord = sentence1[...]

#### Exercise 2
Suppose you have a string `name` containing the name of a Python script. Use slicing/indexing to select the file name without the extension `.py`. Replace the dots between the brackets:

In [ ]:
name = "intro_to_Python.py"
filename = name[...]

## 2.5 String formatting with f-strings

With the use of `f` strings variables (containing integers, floats, strings, ...) can be put together in a formatted string. Formatting can be 
* the number op characters (places) to be used
* desired precision (number of decimals)
* alignment specification: left, right, center

#### Basic formatting

To use `f` strings:
* place an `f` or an `F` before the string to be formatted
* type accolades `{}` (_place holders_) around the variable to be formatted
* let the variable be followed by a colon `:`
* specify the datatype of the variable:

| Character | Datatype |
| :---: | --- |
| `s` | string |
| `d` | integer |
| `f` | float |

* print the `f` string with the `print()` function

The syntax to format a variable is

```Python
    {variable:xtype}
```

with:
* `variable`: the name of the variable to be formatted
* `x`: the number of characters (places) to be used
* `type`: the datatype of the variable to be formatted (`s`, `d` or `f`)

#### Specifying the number of characters

To specify the number of characters to be used, simply `x` by this number:

In [ ]:
char = "a"
n = 23
print(f"The character {char:5s} occurs {n:10d} times")

Observe that:
* since **strings** are aligned **left**, the character `a` is **followed** by 4 additional spaces
* since **numbers** are aligned **right**, the number `23` is **preceded** by 8 additional spaces

#### Precision

In case of decimal numbers (floats), one can specify the precision. This can be realized by using the following syntax:

```Python
    {variable:x.yf}
```

with:
* `variable`: the name of the variable to be formatted
* `x`: the number of characters (places) to be used
* `y`: the number of decimal places to be used

Execute the following code and try to interpret the result:

In [ ]:
bmi = 21.718
name = "John"
print(f"{name:10s} has a BMI of {bmi:7.1f}.")

The formatting `{name:10s}` means: use 10 characters to represent the value of `name`. Since 10 places were provided and strings are aligned to the left, and `John` takes 4 places, 6 additional spaces were added to the **right** of `John`. The `s` stands for string.

The formatting `{bmi:7.1f}` means: use 7 characters to represent the value of `bmi` and round to 1 decimal place. The `f` stands for float.

Since 7 places were provided, and numbers are aligned to the right, and `21.7` takes 4 places (the dot takes one too), 3 additional spaces were added to the **left** of `21.7`. 

#### Alignment
The alignment can be specified by adding `<`, `>` or `^` just after the colon:

| Character | Alignment |
| :---: | --- |
| `<` | left |
| `>` | right |
| `^` | centered |

< [1 The Basics of Python](1-TheBasics.ipynb) | [Contents](0-Contents.ipynb) | [3 Control Structures](3-ControlStructures.ipynb) >